File: packages/filters-gl/src/medianFilter.ts
import type { GlRenderTarget } from '@flighthq/render-gl';
import { compileGlFullscreenProgram, drawGlFullscreenPass } from '@flighthq/render-gl';
import type { MedianFilter } from '@flighthq/types';
import type { GlFullscreenProgram, GlRenderState } from '@flighthq/types';

// Supports radius up to 2 (5×5 = 25 samples). For larger radii use the
// surface path. Sorts per-channel independently using insertion sort.
const MAX_RADIUS = 2;
const MAX_SAMPLES = (MAX_RADIUS * 2 + 1) * (MAX_RADIUS * 2 + 1); // 25

const MEDIAN_FRAGMENT_SRC = `#version 300 es
precision mediump float;
in vec2 v_texCoord;
uniform sampler2D u_texture;
uniform vec2 u_texelSize;
uniform int u_radius;
out vec4 fragColor;

const int MAX_S = ${MAX_SAMPLES};

void sortFloat(inout float arr[MAX_S], int n) {
  for (int i = 1; i < n; i++) {
    float key = arr[i];
    int j = i - 1;
    while (j >= 0 && arr[j] > key) {
      arr[j + 1] = arr[j];
      j--;
    }
    arr[j + 1] = key;
  }
}

void main() {
  int r = clamp(u_radius, 0, ${MAX_RADIUS});
  if (r == 0) {
    fragColor = texture(u_texture, v_texCoord);
    return;
  }
  int n = (2 * r + 1) * (2 * r + 1);
  float rv[MAX_S];
  float gv[MAX_S];
  float bv[MAX_S];
  float av[MAX_S];
  int count = 0;
  for (int dy = -${MAX_RADIUS}; dy <= ${MAX_RADIUS}; dy++) {
    for (int dx = -${MAX_RADIUS}; dx <= ${MAX_RADIUS}; dx++) {
      if (abs(dy) <= r && abs(dx) <= r) {
        vec4 s = texture(u_texture, v_texCoord + vec2(float(dx), float(dy)) * u_texelSize);
        rv[count] = s.r;
        gv[count] = s.g;
        bv[count] = s.b;
        av[count] = s.a;
        count++;
      }
    }
  }
  sortFloat(rv, n);
  sortFloat(gv, n);
  sortFloat(bv, n);
  sortFloat(av, n);
  int mid = n / 2;
  fragColor = vec4(rv[mid], gv[mid], bv[mid], av[mid]);
}`;

type MedianShaderLocations = GlFullscreenProgram & {
  locTexelSize: WebGLUniformLocation;
  locRadius: WebGLUniformLocation;
};

const shaders = new WeakMap<GlRenderState, MedianShaderLocations>();

/**
 * Applies a per-channel median filter to `source`, writing to `dest`.
 * Preserves edges while removing noise. Supports radius 0–2 (up to 5×5);
 * use `applyMedianFilterToSurface` for larger radii.
 * A single GPU pass — no scratch targets needed.
 */
export function applyMedianFilterToGl(
  state: GlRenderState,
  source: GlRenderTarget,
  dest: GlRenderTarget,
  filter: Readonly<Omit<MedianFilter, 'type'>>,
): void {
  const radius = Math.min(MAX_RADIUS, Math.max(0, Math.round(filter.radius ?? 1)));
  const loc = getShader(state);
  drawGlFullscreenPass(state, loc, [source.texture], dest, (gl) => {
    gl.uniform2f(loc.locTexelSize, 1 / source.width, 1 / source.height);
    gl.uniform1i(loc.locRadius, radius);
  });
}

function getShader(state: GlRenderState): MedianShaderLocations {
  let loc = shaders.get(state);
  if (loc === undefined) {
    const gl = state.gl;
    const base = compileGlFullscreenProgram(gl, MEDIAN_FRAGMENT_SRC);
    loc = {
      ...base,
      locTexelSize: gl.getUniformLocation(base.program, 'u_texelSize')!,
      locRadius: gl.getUniformLocation(base.program, 'u_radius')!,
    };
    shaders.set(state, loc);
  }
  return loc;
}

File: crates/flighthq-filters-gl/src/median_filter.rs
//! Median GL filter pass.
//!
//! Preserves edges while removing noise. Supports radius 0–2 (up to 5×5 kernel);
//! use `apply_median_filter_to_surface` from `flighthq-filters` for larger radii.
//! A single GPU pass — no scratch targets needed.

use flighthq_filters::MedianFilter;
use glow::HasContext;

use crate::filter_pass::{GlFullscreenProgram, draw_gl_fullscreen_pass, get_gl_filter_program};
use crate::{GlRenderState, GlRenderTarget};

// Supports radius up to 2 (5×5 = 25 samples); larger radii use the surface path.
// Sorts each channel independently with insertion sort.
const MAX_RADIUS: i32 = 2;

const MEDIAN_FRAGMENT_SRC: &str = "#version 300 es
precision mediump float;
in vec2 v_texCoord;
uniform sampler2D u_texture;
uniform vec2 u_texelSize;
uniform int u_radius;
out vec4 fragColor;

const int MAX_S = 25;

void sortFloat(inout float arr[MAX_S], int n) {
  for (int i = 1; i < n; i++) {
    float key = arr[i];
    int j = i - 1;
    while (j >= 0 && arr[j] > key) {
      arr[j + 1] = arr[j];
      j--;
    }
    arr[j + 1] = key;
  }
}

void main() {
  int r = clamp(u_radius, 0, 2);
  if (r == 0) {
    fragColor = texture(u_texture, v_texCoord);
    return;
  }
  int n = (2 * r + 1) * (2 * r + 1);
  float rv[MAX_S];
  float gv[MAX_S];
  float bv[MAX_S];
  float av[MAX_S];
  int count = 0;
  for (int dy = -2; dy <= 2; dy++) {
    for (int dx = -2; dx <= 2; dx++) {
      if (abs(dy) <= r && abs(dx) <= r) {
        vec4 s = texture(u_texture, v_texCoord + vec2(float(dx), float(dy)) * u_texelSize);
        rv[count] = s.r;
        gv[count] = s.g;
        bv[count] = s.b;
        av[count] = s.a;
        count++;
      }
    }
  }
  sortFloat(rv, n);
  sortFloat(gv, n);
  sortFloat(bv, n);
  sortFloat(av, n);
  int mid = n / 2;
  fragColor = vec4(rv[mid], gv[mid], bv[mid], av[mid]);
}";

/// Applies a per-channel median filter to `source`, writing to `dest`.
pub fn apply_median_filter_to_gl(
    state: &GlRenderState,
    source: &GlRenderTarget,
    dest: &GlRenderTarget,
    filter: &MedianFilter,
) {
    let radius = (filter.radius.unwrap_or(1.0).round() as i32).clamp(0, MAX_RADIUS);
    let (tx, ty) = (1.0 / source.width as f32, 1.0 / source.height as f32);
    let program = get_median_shader(state);
    draw_gl_fullscreen_pass(
        state,
        program,
        &[source.texture],
        Some(dest),
        move |gl, p| unsafe {
            gl.uniform_2_f32(gl.get_uniform_location(p, "u_texelSize").as_ref(), tx, ty);
            gl.uniform_1_i32(gl.get_uniform_location(p, "u_radius").as_ref(), radius);
        },
    );
}

/// Returns the median shader program for `state`, compiling on first use.
pub fn get_median_shader(state: &GlRenderState) -> &GlFullscreenProgram {
    get_gl_filter_program(state, MEDIAN_FRAGMENT_SRC, |p| &mut p.median)
}

#[cfg(test)]
mod tests {
    use super::*;

    // MEDIAN_FRAGMENT_SRC

    #[test]
    fn median_fragment_src_sorts_each_channel() {
        assert!(MEDIAN_FRAGMENT_SRC.contains("void sortFloat"));
        assert!(
            MEDIAN_FRAGMENT_SRC.contains("fragColor = vec4(rv[mid], gv[mid], bv[mid], av[mid])")
        );
    }
}
